File: leon-core/src/test/resources/io/leon/javascript/test/testObjects.js
leon.utils.createVar("io.leon.javascript.test.Tests");

io.leon.javascript.test.Tests = (function() {

  var service = TestService;

  return {
    testGetBean: function() {
      var bean = service.getTestBean();

      if(bean.x() != "x") throw "bean.x is " + bean.x + " but expected 'x'";
      if(bean.y() != 1) throw "bean.y is " + bean.x + " but expected '1'";
    },

    testSetBean: function() {
      service.setTestBean({
        "x": "Hello World",
        "y": 29322,
        "z": { "a": "abcdef", "b": 12389384 },
        "xs": ["a", "b", "c"]
      });
    },

    testMethodWithNumericArgs: function() {
      var x = service.methodWithNumericArgs(10, 10, 0, 1, 0.5, 1.5);
      if(x != 23) throw "exepcted 23 but got " + x;
    },

    testMethodWithString: function() {
      var x = service.methodWithString("Just call me Scala");
      if(x != "Hello World") throw "expected 'Hello World' but got: " + x;
    },

    testMethodWithJavaType: function() {
      var bean = service.getTestBean();
      service.setTestBean(bean);
    },

    testApplyMethodCall: function() {
      var data = {
        "x": "Hello World",
        "y": 29322,
        "z": { "a": "abcdef", "b": 12389384 }
      };

      service.setTestBean.apply(service, [data]);
    },

    methodWithJavaTestBean: function() {
      service.methodWithJavaTestBean({
        "x": "test",
        "y": 1,
        "z": { "x": "test" },
        "xs": ["a", "b", "c"]
      }).toJSON();
    },

    methodWithJavaList: function() {
      var bean = {
        "x": "test",
        "y": 1,
        "z": { "x": "test" }
      };

      var list = [bean, bean, bean];
      var result = service.methodWithJavaList(list).toJSON();

      if(result.length != list.length) throw "expected " + list.length + " but got " + result.length;

      for(var i = 0; i < result.length; i++)
        if(result[i].x != "test") throw "expected 'test' but got: " + result[i].x;
    },

    methodWithSeq: function() {
      var bean = {
        "x": "Hello World",
        "y": 1,
        "z": { "a": "abcdef", "b": 12389384 }
      };

      var list = [bean, bean, bean]
      var result = service.methodWithSeq(list).toJSON();

      if(result.length != list.length) throw "expected " + list.length + " but got " + result.length;

      for(var i = 0; i < result.length; i++)
        if(result[i].y != (bean.y + 1)) throw "expected " + (bean.y + 1) + " but got: " + result[i].y;
    },

    methodWithArray: function() {
      var result = service.methodWithArray([1, 2, 3]);
      if(result.length != 3) throw "expected 3 but got " + result.length;

      for(var i=0; i < result.length; i++)
        if(result[i] != (i + 2)) throw "expected " + (i + 2) + " but got " + result[i];
    },

    asJavaObject: function() {
      var bean = {
        "x": "test",
        "y": 1,
        "z": { "x": "test" }
      };

      bean.asJavaObject(Packages.io.leon.javascript.test.TestBean);
    }
  };
})();

File: leon-core/src/test/scala/io/leon/javascript/test/testObjects.scala
package io.leon.javascript.test

import reflect.BeanInfo
import java.lang.{Float, Short}


@BeanInfo
case class TestBean(x: String, y: Int, z: NestedTestBean, xs: Array[String]) {
  private def this() = this("", 0, null, null)
}

@BeanInfo
case class NestedTestBean(a: String, b: Long) {
  private def this() = this("", 0)
}


class TestService {

  private var bean = TestBean("x", 1, NestedTestBean("a", 2), Array("a"))

  def getTestBean = bean

  def setTestBean(bean: TestBean) {
    this.bean = bean
  }

  def methodWithNumericArgs(i: Int, l: Long, s: Short, b: Byte, f: Float, d: Double) = {
    (i + l + s + b + f + d)
  }

  def methodWithDefaultArgs(bean: TestBean, x: Int = 2) {

  }

  def methodWithString(s: String) = {
    "Hello World"
  }

  def methodWithList(list: Seq[TestBean]) {
  }

  def methodWithIntList(list: Seq[Int]) {
  }

  def methodWithJavaTestBean(bean: JavaTestBean) = {
    bean.setX("Hello world")
    bean
  }

  def methodWithJavaList(list: java.util.List[JavaTestBean]) = {
    list
  }

  def methodWithSeq(seq: Seq[TestBean]) = {
    seq map { e => e.copy(y = e.y + 1) }
  }

  def methodWithArray(arr: Array[Int]) = {
    arr map { _ + 1 }
  }

  def overloadedMethod(i: Int, bean: TestBean) {}

  def overloadedMethod(bean: TestBean, i: Int) {}


}
